### PR TITLE
Add sitter onboarding view and view model

### LINKED ISSUE #99: Add `SitterOnboardingView` and viewModel
## This task is a part of the issue:
 - #96 

## Depends on tasks:
- #_issue
- #_issue

### PR DESCRIPTION
**Pull request by iBYarchitect**
## Summary
This PR is a part of the work to resolve #99

## UI Preview (Screenshots)
- [x] iPhone 15 Pro -  6.1" 
image

## Changes
- Added `SitterOnboardingView`
- Added `SitterOnboardingViewModel`
- Changed `EditSitterInformationView`

## Checks to complete
- [x] The PR title has a conventional commit style (short and descriptive).
- [x] The PR description includes essential information that helps understand the purpose of the changes.
- [x] I've built and run my changes, and no warnings or errors occurred.
- [x] All existing tests passed with my changes.
- [x] My code follows our style guide.
- [x] I've performed a self-review of my code.
- [x] Every non-private interface is documented properly.
- [x] I've added tests for my code (if it makes sense).
- [x] I've updated `CHANGELOG.md` file.
- [x] PR has assignee, reviewers, milestone, it is properly linked to the project and to the issue.

### CODE REVIEW

File: Woof/Woof/View/SitterFlow/SitterProfileView/EditSitterInformationView.swift [hunk 1]
```diff
@@ -45,6 +46,9 @@ struct EditSitterInformationView: View {
                 .keyboardType(.decimalPad)
             TextEditorWithPlaceholder(text: $bio, placeholderText: bioInformationPlaceholderText)
                 .cornerRadius(AppStyle.UIElementConstant.cornerRadius)
+            if name.isEmpty || phone.isEmpty || city.isEmpty || pricePerHour.isEmpty {
+                Text(mandatoryText)
+            }
```
Comment (NadzeyaShpakouskaya) [inline Woof/Woof/View/SitterFlow/SitterProfileView/EditSitterInformationView.swift:51]: Why do we need this? We have already had this info in the profile view.
Simulator Screenshot - iPhone 15 - 2024-03-14 at 11 04 16

**Review by NadzeyaShpakouskaya — CHANGES_REQUESTED** [review]
Please consider some improvements:

File: Woof/Woof/View/SitterFlow/SitterOnboardingView.swift [hunk 2]
```diff
@@ -0,0 +1,53 @@
+import SwiftUI
+
+/// A view to onboard the sitter by filling in the mandatory information about themselves.
+struct SitterOnboardingView: View {
+    /// View model responsible to manage data from model layer
```
Comment (kirpichonok) [inline Woof/Woof/View/SitterFlow/SitterOnboardingView.swift:5]: What `SitterProfileViewModel`'s functionality do you want to change? Or what functionality do you want to add? 
```suggestion
struct SitterOnboardingView: View {
    typealias SitterOnboardingViewModel = SitterProfileViewModel
    /// View model responsible to manage data from model layer
```

File: Woof/Woof/View/SitterFlow/SitterOnboardingView.swift [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+                .buttonStyle(CapsuleWithWhiteText())
+                .padding()
+                .disabled(isSaveButtonEnabled)
+        }
+        .padding()
+    }
+
+    // MARK: - Private interface
+
+    private let welcomeText = "Hello!"
```
Comment (kirpichonok) [inline Woof/Woof/View/SitterFlow/SitterOnboardingView.swift:43]: nit: Just try that )
```suggestion
    private let welcomeText = "Hello!👋🏻!"
```

File: Woof/Woof/View/SitterFlow/SitterOnboardingView.swift [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    private let proceedButtonTitle = "Start earn money!"
+    private let onboardingText = """
+    We're excited to have you on board as a pet sitter!
+    Your dedication and care will make a world of difference to pets in need. Let's begin by entering your information:
+    """
```
Comment (kirpichonok) [inline Woof/Woof/View/SitterFlow/SitterOnboardingView.swift:48]: It looks superfluous and doesn't match with "Start earning money!"
```suggestion
    Let's begin by entering your information:
    """
```

File: Woof/Woof/ViewModel/Sitter/SitterOnboardingViewModel.swift
Comment (kirpichonok) [inline Woof/Woof/ViewModel/Sitter/SitterOnboardingViewModel.swift:1]: I don't seem fully to understand how you intend to use it.

**Review by kirpichonok — CHANGES_REQUESTED** [review]
I have some serious questions about your implementation. We can get in touch and discuss it.

File: Woof/Woof/View/SitterFlow/SitterOnboardingView.swift (hunk 4 above)
Comment (iBYarchitect) [reply Woof/Woof/View/SitterFlow/SitterOnboardingView.swift]: Fixed with 5e32277

File: Woof/Woof/ViewModel/Sitter/SitterOnboardingViewModel.swift
Comment (iBYarchitect) [reply Woof/Woof/ViewModel/Sitter/SitterOnboardingViewModel.swift:1]: Yes, you're right, this VM is not needed.
Deleted with 5e32277

File: Woof/Woof/View/SitterFlow/SitterOnboardingView.swift (hunk 3 above)
Comment (iBYarchitect) [reply Woof/Woof/View/SitterFlow/SitterOnboardingView.swift]: Added with 5e32277

File: Woof/Woof/View/SitterFlow/SitterOnboardingView.swift [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+            EditSitterInformationView(
+                name: $viewModel.name,
+                surname: $viewModel.surname,
+                phone: $viewModel.phone,
+                city: $viewModel.city,
+                bio: $viewModel.bio,
+                pricePerHour: $viewModel.pricePerHour
+            )
+
+            Button(proceedButtonTitle) {}
```
[the remark this answers is not in the document]
Comment (iBYarchitect) [reply Woof/Woof/View/SitterFlow/SitterOnboardingView.swift]: Added with 5e32277

File: Woof/Woof/View/SitterFlow/SitterOnboardingView.swift [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+            viewModel.phone.isEmpty ||
+            viewModel.city.isEmpty ||
+            viewModel.pricePerHour.isEmpty
+    }
+
+    var body: some View {
+        VStack {
+            Text(welcomeText)
+                .foregroundColor(Color.App.purpleDark)
+
```
[the remark this answers is not in the document]
Comment (iBYarchitect) [reply Woof/Woof/View/SitterFlow/SitterOnboardingView.swift]: Added with 5e32277

File: Woof/Woof/View/SitterFlow/SitterOnboardingView.swift (hunk 2 above)
Comment (iBYarchitect) [reply Woof/Woof/View/SitterFlow/SitterOnboardingView.swift:5]: Yes, you're right, the functionality of the current view model will be used, and nothing new is needed.

File: Woof/Woof/View/SitterFlow/SitterProfileView/EditSitterInformationView.swift (hunk 1 above)
Comment (iBYarchitect) [reply Woof/Woof/View/SitterFlow/SitterProfileView/EditSitterInformationView.swift]: Fixed with 5e32277

File: Woof/Woof/View/SitterFlow/SitterOnboardingView.swift [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+                phone: $viewModel.phone,
+                city: $viewModel.city,
+                bio: $viewModel.bio,
+                pricePerHour: $viewModel.pricePerHour
+            )
+
+            Button(proceedButtonTitle) {}
+                .buttonStyle(CapsuleWithWhiteText())
+                .padding()
+                .disabled(isSaveButtonEnabled)
```
[the remark this answers is not in the document]
Comment (iBYarchitect) [reply Woof/Woof/View/SitterFlow/SitterOnboardingView.swift]: Fixed with 5e32277

File: Woof/Woof/View/SitterFlow/SitterOnboardingView.swift [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+    /// View model responsible to manage data from model layer
+    @ObservedObject var viewModel = SitterOnboardingViewModel()
+
+    /// A boolean indicating whether the save button should be enabled.
+    var isSaveButtonEnabled: Bool {
+        viewModel.name.isEmpty ||
+            viewModel.phone.isEmpty ||
+            viewModel.city.isEmpty ||
+            viewModel.pricePerHour.isEmpty
+    }
```
[the remark this answers is not in the document]
Comment (iBYarchitect) [reply Woof/Woof/View/SitterFlow/SitterOnboardingView.swift]: I used `mandatoryFieldsAreEmpty` in 5e32277

File: Woof/Woof/View/SitterFlow/SitterOnboardingView.swift [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+                .padding()
+                .disabled(isSaveButtonEnabled)
+        }
+        .padding()
+    }
+
+    // MARK: - Private interface
+
+    private let welcomeText = "Hello!"
+    private let proceedButtonTitle = "Start earn money!"
```
[the remark this answers is not in the document]
Comment (iBYarchitect) [reply Woof/Woof/View/SitterFlow/SitterOnboardingView.swift]: Fixed in 5e32277

File: Woof/Woof/View/SitterFlow/SitterOnboardingView.swift [hunk 10]
```diff
@@ -0,0 +1,53 @@
+import SwiftUI
+
+/// A view to onboard the sitter by filling in the mandatory information about themselves.
+struct SitterOnboardingView: View {
+    /// View model responsible to manage data from model layer
+    @ObservedObject var viewModel = SitterOnboardingViewModel()
```
[the remark this answers is not in the document]
Comment (iBYarchitect) [reply Woof/Woof/View/SitterFlow/SitterOnboardingView.swift]: Fixed with 5e32277

[the remark this answers is not in the document]
Comment (NadzeyaShpakouskaya) [reply Woof/Woof/View/SitterFlow/SitterOnboardingView.swift]: you forgot the private annotation, we don't net expose viewModel.

[the remark this answers is not in the document]
Comment (NadzeyaShpakouskaya) [reply Woof/Woof/View/SitterFlow/SitterOnboardingView.swift]: we can mark it private.

File: Woof/Woof/View/SitterFlow/SitterOnboardingView.swift [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+            .padding()
+            .disabled(viewModel.mandatoryFieldsAreEmpty)
+        }
+        .padding()
+        .overlay(
+            Group {
+                if viewModel.isSavingData {
+                    Color.white.opacity(AppStyle.UIElementConstant.opacityLevelForProgressViewBackground)
+                    ProgressView()
+                        .progressViewStyle(CircularProgressViewStyle())
```
Comment (NadzeyaShpakouskaya) [inline Woof/Woof/View/SitterFlow/SitterOnboardingView.swift:49]: Does it make sense to extract this into a separate view? For now, we use it in the onboarding view and profile view. But I'm sure we will reuse it for owner flow as well.

File: Woof/Woof/ViewModel/Sitter/SitterProfileViewModel.swift [hunk 12]
```diff
@@ -3,7 +3,7 @@ import NetworkService
 
 /// The view model for a pet sitter profile view is responsible for receiving and processing user actions
 /// to change model data and passes back the updated model data.
-final class SitterProfileViewModel: ObservableObject {
+class SitterProfileViewModel: ObservableObject {
```
Comment (NadzeyaShpakouskaya) [inline Woof/Woof/ViewModel/Sitter/SitterProfileViewModel.swift:6]: make it final back, as we do not inherit or extend it now.

**Review by NadzeyaShpakouskaya — COMMENTED** [review]
Please check some comments below:

File: Woof/Woof/View/SitterFlow/SitterOnboardingView.swift (hunk 2 above)
Comment (kirpichonok) [reply Woof/Woof/View/SitterFlow/SitterOnboardingView.swift]: WDYT about to add type alias in my suggestion above:
`typealias SitterOnboardingViewModel = SitterProfileViewModel`

File: Woof/Woof/View/SitterFlow/SitterOnboardingView.swift (hunk 11 above)
Comment (iBYarchitect) [reply Woof/Woof/View/SitterFlow/SitterOnboardingView.swift]: Created `CustomProgressView` with 4488b64

File: Woof/Woof/View/SitterFlow/SitterOnboardingView.swift [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+
+/// A view to onboard the sitter by filling in the mandatory information about themselves.
+struct SitterOnboardingView: View {
+    /// View model responsible to manage data from model layer
+    @StateObject var viewModel = SitterProfileViewModel()
+
+    var body: some View {
+        VStack {
+            Text(welcomeText)
+                .foregroundColor(Color.App.purpleDark)
```
[the remark this answers is not in the document]
Comment (iBYarchitect) [reply Woof/Woof/View/SitterFlow/SitterOnboardingView.swift]: Fixed with 4488b64

File: Woof/Woof/ViewModel/Sitter/SitterProfileViewModel.swift (hunk 12 above)
Comment (iBYarchitect) [reply Woof/Woof/ViewModel/Sitter/SitterProfileViewModel.swift]: Returned with 4488b64

File: Woof/Woof/View/SitterFlow/SitterOnboardingView.swift (hunk 5 above)
[the remark this answers is not in the document]
Comment (kirpichonok) [reply Woof/Woof/View/SitterFlow/SitterOnboardingView.swift]: Your implementation doesn't consider errors during saving process.

[the remark this answers is not in the document]
Comment (iBYarchitect) [reply Woof/Woof/View/SitterFlow/SitterOnboardingView.swift]: Please look at the code below, `.alert` part

File: Woof/Woof/View/SitterFlow/SitterOnboardingView.swift (hunk 10 above)
[the remark this answers is not in the document]
Comment (iBYarchitect) [reply Woof/Woof/View/SitterFlow/SitterOnboardingView.swift]: fixed with 4488b64

File: Woof/Woof/View/SitterFlow/SitterOnboardingView.swift (hunk 11 above)
Comment (NadzeyaShpakouskaya) [reply Woof/Woof/View/SitterFlow/SitterOnboardingView.swift]: I meant to  extract a white transparent background and progress view, it doesn't make sense to extract only the progress view

File: Woof/Woof/View/SitterFlow/SitterOnboardingView.swift [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+            }
+            .buttonStyle(CapsuleWithWhiteText())
+            .padding()
+            .disabled(viewModel.mandatoryFieldsAreEmpty)
+        }
+        .foregroundColor(Color.App.purpleDark)
+        .padding()
+        .overlay {
+            if viewModel.isSavingData {
+                CustomProgressView()
```
[the remark this answers is not in the document]
Comment (iBYarchitect) [reply Woof/Woof/View/SitterFlow/SitterOnboardingView.swift:41]: we're following DRY principle

File: Woof/Woof/View/ReusableViews/CustomProgressView.swift [hunk 15]
```diff
@@ -0,0 +1,14 @@
+import SwiftUI
+
+/// /// A view representing a custom circular progress indicator.
```
Comment (kirpichonok) [inline Woof/Woof/View/ReusableViews/CustomProgressView.swift:3]: Fix it.
```suggestion
/// A view representing a custom circular progress indicator.
```

File: Woof/Woof/View/ReusableViews/CustomProgressView.swift [hunk 16]
```diff
@@ -0,0 +1,14 @@
...
+struct CustomProgressView: View {
+    var body: some View {
+        Color.white.opacity(AppStyle.UIElementConstant.opacityLevelForProgressViewBackground)
+        ProgressView()
+            .progressViewStyle(CircularProgressViewStyle())
```
Comment (kirpichonok) [inline Woof/Woof/View/ReusableViews/CustomProgressView.swift:8]: WDYT?
```suggestion
        ZStack {
            Color.white.opacity(
                AppStyle.UIElementConstant.opacityLevelForProgressViewBackground
            )
            ProgressView()
        }
```

**Review by kirpichonok — CHANGES_REQUESTED** [review]
Fix a documentation and consider one suggestion.

File: Woof/Woof/View/SitterFlow/SitterOnboardingView.swift (hunk 11 above)
Comment (NadzeyaShpakouskaya) [reply Woof/Woof/View/SitterFlow/SitterOnboardingView.swift]: Could you replace it with the new view in the changed code in `SitterProfileView`as well?

**Review by NadzeyaShpakouskaya — APPROVED** [review]
LGTM after resolving all conversation

File: Woof/Woof/View/SitterFlow/SitterOnboardingView.swift (hunk 11 above)
Comment (iBYarchitect) [reply Woof/Woof/View/SitterFlow/SitterOnboardingView.swift]: Yes, updated with 2a4848e

File: Woof/Woof/View/ReusableViews/CustomProgressView.swift (hunk 15 above)
Comment (iBYarchitect) [reply Woof/Woof/View/ReusableViews/CustomProgressView.swift]: Fixed with 2a4848e

File: Woof/Woof/View/ReusableViews/CustomProgressView.swift (hunk 16 above)
Comment (iBYarchitect) [reply Woof/Woof/View/ReusableViews/CustomProgressView.swift]: Reworked with 9fad9ce

Comment (kirpichonok) [inline Woof/Woof/View/ReusableViews/CustomProgressView.swift]: Why do you need this line?

**Review by kirpichonok — APPROVED** [review]
LGTM

**Pull request merged**